radiosity.cpp

```diff
@@ -0,0 +1,327 @@
+        assert(lightNode != nullptr);
+        const auto& LT = lightNode->mTransformation;
+        lights.push_back(
+            { LT * aiVector3D()
+            , aiColor4D
+                { rawLight.mColorDiffuse.r
+                , rawLight.mColorDiffuse.g
+                , rawLight.mColorDiffuse.b
+                , 1
+                }
+            });
+    }
+
+    // load triangles from the scene into a kd-tree
+    auto triangles = triangles_from_scene(scene);
+    Stats::instance().num_triangles = triangles.size();
+    Tree tree(std::move(triangles));
+
+    //
+    // Raytracer
+    //
+
+    int width = args["--width"].asLong();
+    assert(width > 0);
+    int height = width / cam.mAspect;
+
+    Image image(width, height);
+    {
+        Runtime rt(Stats::instance().runtime_ms);
+
+        std::cerr << "Rendering          ";
+
+        ThreadPool pool(conf.num_threads);
+        std::vector<std::future<void>> tasks;
+
+        for (int y = 0; y < height; ++y) {
+            tasks.emplace_back(pool.enqueue([
+                    &image, &cam, &tree, &lights, width, height, y, &conf]()
+            {
+                float dx, dy;
+                xorshift64star<float> gen(42);
+
+                for (int x = 0; x < width; ++x) {
+                    for (int i = 0; i < conf.num_pixel_samples; ++i) {
+                        dx = gen();
+                        dy = gen();
+
+                        auto cam_dir = cam.raster2cam(
+                            aiVector2D(x + dx, y + dy), width, height);
+
+                        Stats::instance().num_prim_rays += 1;
+                        image(x, y) += trace(cam.mPosition, cam_dir, tree,
+                            lights, 0, conf);
+                    }
+                    image(x, y) /=
+                        static_cast<float>(conf.num_pixel_samples);
+
+                    // gamma correction
+                    if (conf.gamma_correction_enabled) {
+                        image(x, y).r = powf(image(x, y).r, conf.inverse_gamma);
+                        image(x, y).g = powf(image(x, y).g, conf.inverse_gamma);
+                        image(x, y).b = powf(image(x, y).b, conf.inverse_gamma);
+                    }
+                }
+            }));
+        }
+
+        long completed = 0;
+
+        for (auto& task: tasks) {
+            task.get();
+            completed += 1;
+            float progress = static_cast<float>(completed) / tasks.size();
+            int bar_width = progress * 20;
+            std::cerr
+                << "\rRendering          "
+                << "[" << std::string(bar_width, '-')
+                << std::string(20 - bar_width, ' ') << "] "
+                << std::setfill(' ') << std::setw(6)
+                << std::fixed << std::setprecision(2) << (progress * 100.0) << '%';
+            std::cerr.flush();
+        }
+        std::cerr << std::endl;
+
+        // Render feature lines after
```

Comment:
Könnte man vielleicht diesen Part als Injection in die Original-Main reinstecken?